webrender/src/device/gfx/descriptor.rs

```diff
@@ -2,409 +2,519 @@
-            ShaderGroup::ClipCache => {
-                assert!(location.pool_idx <= self.cache_clip_pool_idx);
-                &self.cache_clip_pool[location.pool_idx]
-            },
-            ShaderGroup::Brush => {
-                assert!(location.pool_idx <= self.brush_pool_idx);
-                &self.brush_pool[location.pool_idx]
-            },
-        };
-        pool.descriptor_set_at_idx(location.set_idx)
+    pub(super) fn retain(&mut self, id: &TextureId) {
+        let keys_to_remove: Vec<_> = self.descriptor_bindings
+            .keys()
+            .filter(|res| res.has_texture_id(&id)).cloned().collect();
+        for key in keys_to_remove {
+            let desc_set =self.descriptor_bindings.remove(&key).unwrap();
+            self.free_sets.get_mut(key.desc_group()).push(desc_set);
+        }
     }
 
-    pub(super) fn get_layout(&self, shader_group: ShaderGroup) -> &B::DescriptorSetLayout {
-        self.get_pool(shader_group).descriptor_set_layout()
+    pub(super) unsafe fn free(self, allocator: &mut DescriptorAllocator<B>) {
+        self.free_sets.free(allocator);
+        allocator.free(self.descriptor_bindings.into_iter().map(|(_, set)| set));
     }
 
-    pub(super) fn mark_as_free(&mut self, shader_group: ShaderGroup, location: DescriptorSetLocation) {
-        let pool = match shader_group {
-            ShaderGroup::Debug => &mut self.debug_pool,
-            ShaderGroup::ClipCache => {
-                if location.pool_idx != self.cache_clip_pool_idx
-                    && !self.marked_cache_clip_pools.contains(&location.pool_idx) {
-                    self.marked_cache_clip_pools.push(location.pool_idx)
+    pub(super) fn bind_textures(
+        &mut self,
+        bound_textures: &[u32; RENDERER_TEXTURE_COUNT],
+        cmd_buffer: &mut hal::command::CommandBuffer<B, hal::Graphics>,
+        bindings: K,
+        images: &FastHashMap<TextureId, Image<B>>,
+        desc_allocator: &mut DescriptorAllocator<B>,
+        device: &B::Device,
+        group_data: &DescriptorGroupData<B>,
+        group: &DescriptorGroup,
+        set_index: usize,
+        range: std::ops::Range<usize>,
+        sampler: Option<&B::Sampler>,
+    ) {
+        match self.descriptor_bindings.entry(bindings) {
+            Entry::Occupied(_) => {
+                for index in range {
+                    let image = &images[&bound_textures[index]].core;
+                    // We need to transit the image, even though it's bound in the descriptor set
+                    let mut src_stage = Some(hal::pso::PipelineStage::empty());
+                    if let Some(barrier) = image.transit(
+                        hal::image::Access::SHADER_READ,
+                        hal::image::Layout::ShaderReadOnlyOptimal,
+                        image.subresource_range.clone(),
+                        src_stage.as_mut(),
+                    ) {
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
```

Comment:
It would be good to combine all the transitions into a single `pipeline_barrier` call.
Like with any "would be good" things, please either address it in the PR, or add a descriptive TODO comment here explaining what remains to be done.